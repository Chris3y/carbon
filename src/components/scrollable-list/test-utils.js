import React from 'react';
import ScrollableListItem from './scrollable-list-item.component';

const makeArray = n => [...Array(n).keys()];

const listItemReducer = ({ nonSelectables = [], customSelectables = [] }) => {
  // generate jsx for selectable list items, based on indexes in config
  return (acc, item, index) => {
    if (nonSelectables.includes(item)) return [...acc, <div />];
    if (customSelectables.includes(item)) return [...acc, <div isSelectable />];

    return [...acc, <ScrollableListItem id={ index } />];
  };
};

const renderListItems = (opts) => {
  return makeArray(opts.num).reduce(listItemReducer(opts), []);
};


const keyPress = (code) => {
  const ev = new KeyboardEvent('keydown', { which: code });
  document.dispatchEvent(ev);
};

const keyMap = {
  UpArrow: '38',
  DownArrow: '40',
  RightArrow: '39',
  LeftArrow: '37',
  Enter: '13',
  Tab: '9'
};

const repeat = action => (n = 1) => makeArray(n).forEach(() => action());

const keyboard = Object.keys(keyMap).reduce((acc, key) => {
  acc[`press${key}`] = () => repeat(keyPress(keyMap[key]));
  return acc;
}, {});

const listFrom = wrapper => wrapper.find('ul');
const childrenFrom = node => node.children();

const hoverList = wrapper => (item) => {
  childrenFrom(listFrom(wrapper)).at(item).simulate('mouseover');
};

const simulateEvent = eventName => wrapper => wrapper.simulate(eventName);
const click = simulateEvent('click');

const selectedItemOf = wrapper => wrapper.state().selectedItem;
const isUnique = (val, index, self) => self.indexOf(val) === index;
const isSelectableGiven = nonSelectables => i => !nonSelectables.includes(i);

const selectedItemReducer = method => wrapper => (acc, i) => {
  method(wrapper)(i);
  return [...acc, selectedItemOf(wrapper)];
};

const arraysEqual = (arr1, arr2) => arr1.sort().join(',') === arr2.sort().join(',');

const assertCorrectTraversal = method => expect => ({ num, nonSelectables = [] }) => (wrapper) => {
  const array = makeArray(num);
  const validIndexes = array.filter(isSelectableGiven(nonSelectables));

  const indexesThatWereSelected = array
    .reduce(selectedItemReducer(method)(wrapper), [])
    .filter(isUnique);
  expect(arraysEqual(validIndexes, indexesThatWereSelected)).toBeTruthy();
};

const assertKeyboardTraversal = assertCorrectTraversal(() => keyboard.pressDownArrow)(expect);
const assertHoverTraversal = assertCorrectTraversal(wrapper => hoverList(wrapper))(expect);

export {
  hoverList,
  selectedItemOf,
  childrenFrom,
  renderListItems,
  keyboard,
  assertKeyboardTraversal,
  assertHoverTraversal,
  listFrom,
  click
};
