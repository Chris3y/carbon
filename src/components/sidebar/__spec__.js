import React from 'react';
import { shallow, mount } from 'enzyme';
import Browser from '../../utils/helpers/browser';
import { Sidebar } from './sidebar';
import Textbox from './../textbox';
import Portal from './../portal';
import Icon from './../icon';

describe('Sidebar', () => {
  let wrapper, spy;

  beforeEach(() => {
    spy = jasmine.createSpy();

    wrapper = shallow(
      <Sidebar
        open
        title='Test'
        className='custom-class'
        data-role='baz'
        data-element='bar'
        onCancel={ spy }
      >
        <Textbox />
        <Textbox />
        <Textbox />
      </Sidebar>
    );
  });

  describe('componentDidUpdate', () => {
    let mockWindow;
    let wrapper;
    let instance;

    beforeEach(() => {
      mockWindow = {
        addEventListener() {},
        removeEventListener() {}
      };

      spyOn(Browser, 'getWindow').and.returnValue(mockWindow);
    });

    describe('when the Sidebar is open', () => {
      beforeEach(() => {
        wrapper = mount(
          <Sidebar open onCancel={ spy } />
        );
        instance = wrapper.instance();
      });

      it('sets up event listeners to resize and close the Sidebar', () => {
        spyOn(mockWindow, 'addEventListener');
        wrapper.setProps({ title: 'Sidebar title' });

        expect(mockWindow.addEventListener.calls.count()).toEqual(1);
        expect(mockWindow.addEventListener).toHaveBeenCalledWith('keyup', instance.closeModal);
      });

      describe('when the Sidebar is already listening', () => {
        it('does not set up event listeners', () => {
          spyOn(mockWindow, 'addEventListener');
          instance.listening = true;
          wrapper.setProps({ title: 'Already listening' });

          expect(mockWindow.addEventListener.calls.count()).toEqual(0);
          expect(mockWindow.addEventListener).not.toHaveBeenCalled();
        });
      });
    });

    describe('when the Sidebar is closed', () => {
      beforeEach(() => {
        wrapper = mount(
          <Sidebar onCancel={ spy } />
        );
        instance = wrapper.instance();
        instance.listening = true;
      });

      it('removes event listeners for resize and closing', () => {
        spyOn(mockWindow, 'removeEventListener');
        wrapper.setProps({ title: 'Remove event handlers' });
        expect(mockWindow.removeEventListener.calls.count()).toEqual(1);
        expect(mockWindow.removeEventListener).toHaveBeenCalledWith('keyup', instance.closeModal);
      });
    });
  });

  describe('sidebarClasses', () => {
    it('returns a base sidebar', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('render', () => {
    describe('when sidebar is closed', () => {
      it('sets all the correct classes', () => {
        wrapper = mount(<Sidebar onCancel={ spy } />);
        expect(wrapper.find('.carbon-sidebar').text()).toEqual('');
      });
    });

    describe('when enableBackgroundUI is enabled', () => {
      it('sets all the correct classes', () => {
        wrapper = shallow(
          <Sidebar
            open
            enableBackgroundUI
            size='small'
            position='left'
            onCancel={ spy }
          />
        );
        expect(wrapper.find('.carbon-modal__background').length).toEqual(0);
        expect(wrapper.find('.carbon-sidebar__sidebar--left').length).toEqual(1);
        expect(wrapper.find('.carbon-sidebar__sidebar--small').length).toEqual(1);
      });
    });

    describe('when there is no onCancel prop', () => {
      it('should not have a close button', () => {
        wrapper = shallow(
          <Sidebar
            open
          />
        );
        expect(wrapper.find('.carbon-sidebar__close').length).toEqual(0);
      });
    });
  });

  describe('Behaviour', () => {
    describe('clicking the close icon sidebar', () => {
      it('closes the sidebar', () => {
        let icon = wrapper.find('.carbon-sidebar__close-icon');
        icon.simulate('click');
        expect(spy).toHaveBeenCalled();
      });
    });
  });
});
