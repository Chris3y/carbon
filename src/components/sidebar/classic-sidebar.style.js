import { css } from 'styled-components';
import { THEMES } from '../../style/themes';

const sidebarClassicStyle = ({ theme }) => theme.name === THEMES.classic && css`
    background-color: #e6ebed;

     ${({ position }) => position === 'right' && css`
         border-left: 1px solid #ccd6db;
    `};

    ${({ position }) => position === 'left' && css`
        border-right: 1px solid #ccd6db;
    `};
`;

const sidebarCloseClassicStyle = ({ theme }) => theme.name === THEMES.classic && css`
    color: rgba(0, 0, 0, 0.85);

    &:hover {
        color: #255BC7;
    };
`;

export { sidebarClassicStyle, sidebarCloseClassicStyle };
