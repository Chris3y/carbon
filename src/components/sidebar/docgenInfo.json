{
  "src/components/sidebar/sidebar-header/sidebar-header.component.js": [
    {
      "description": "",
      "displayName": "SidebarHeader",
      "methods": [],
      "props": {
        "children": {
          "type": {
            "name": "node"
          },
          "required": false,
          "description": "This component supports children."
        },
        "className": {
          "type": {
            "name": "string"
          },
          "required": false,
          "description": "A custom class name."
        }
      }
    }
  ]
}