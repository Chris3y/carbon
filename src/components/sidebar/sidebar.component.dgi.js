import React from 'react';
import PropTypes from 'prop-types';

class Sidebar extends React.component {
static propTypes = {
  onCancel: PropTypes.func,
  /** A boolean to track the open state of the dialog */
  open: PropTypes.bool,
  /** Set this prop to false to hide the translucent background when the dialog is open. */
  enableBackgroundUI: PropTypes.bool,
  /** Sets the position of sidebar, either left or right. */
  position: PropTypes.string,
  /** Sets the size of the sidebar when open. */
  size: PropTypes.string
}

render() {
  return (<></>);
}
}
export default Sidebar;
