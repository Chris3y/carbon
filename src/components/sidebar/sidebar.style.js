import styled, { css } from 'styled-components';

const sidebarSizes = {
  'extra-small': '150px',
  small: '250px',
  'medium-small': '350px',
  medium: '450px',
  'medium-large': '550px',
  large: '650px',
  'extra-large': '750px'
};

const SidebarStyle = styled.div`
  ${props => console.log(props)}
  background-color: #e6ebed;
  border-radius: 1px;
  bottom: 0;
  overflow: auto;
  padding: 20px;
  position: fixed;
  top: 0;
  z-index: 1002;
  
    ${({ size }) => size && css`
        width: ${sidebarSizes[size]}
    `}

    ${({ position }) => position === 'right' && css`
        box-shadow: -10px 0 15px rgba(0, 0, 0, 0.05);
        border-left: 1px solid #ccd6db;
        right: 0;
    `}

    ${({ position }) => position === 'left' && css`
        box-shadow: 10px 0 15px rgba(0, 0, 0, 0.05);
        border-right: 1px solid #ccd6db;
        left: 0;
    `}

    .carbon-sidebar__close {
        cursor: pointer;
        position: absolute;
        right: 20px;
        top: 15px;
        z-index: 1;

        &:hover {
            color: #255BC7;
        }
    }

`;

export default SidebarStyle;
