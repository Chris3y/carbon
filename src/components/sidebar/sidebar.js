import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import SidebarHeader from './sidebar-header';
import Icon from '../icon';
import Modal from '../modal';
import SidebarStyle from './sidebar.style';
import './sidebar.scss';

class Sidebar extends Modal {
  constructor(props) {
    super(props);
    console.log(props);
  }

  /** Returns classes for the component. */
  get mainClasses() {
    return classNames(
      'carbon-sidebar',
      this.props.className
    );
  }

  /** Returns the markup for the close icon. */
  get closeButton() {
    if (this.props.onCancel) {
      return (
        <span className='carbon-sidebar__close'>
          <Icon
            className='carbon-sidebar__close-icon'
            data-element='close'
            onClick={ this.props.onCancel }
            type='close'
          />
        </span>
      );
    }
    return null;
  }

  componentTags(props) {
    return {
      'data-component': 'sidebar',
      'data-element': props['data-element'],
      'data-role': props['data-role']
    };
  }

  /** Returns the computed HTML for the sidebar. */
  get modalHTML() {
    return (
      <SidebarStyle
        position={ this.props.position }
        size={ this.props.size }
        { ...this.componentTags(this.props) }
      >
        { this.closeButton }
        { this.props.children }
      </SidebarStyle>
    );
  }

  get transitionName() {
    return `sidebar--${this.props.position}`;
  }
}

Sidebar.propTypes = {
  onCancel: PropTypes.func,
  /** A boolean to track the open state of the dialog */
  open: PropTypes.bool,
  /** Set this prop to false to hide the translucent background when the dialog is open. */
  enableBackgroundUI: PropTypes.bool,
  /** Sets the position of sidebar, either left or right. */
  position: PropTypes.string,
  /** Sets the size of the sidebar when open. */
  size: PropTypes.string
};

Sidebar.defaultProps = {
  position: 'right',
  size: 'medium',
  open: false,
  enableBackgroundUI: false
};

export {
  Sidebar,
  SidebarHeader
};
