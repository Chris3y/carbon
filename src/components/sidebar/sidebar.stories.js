import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, select, text } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import { State, Store } from '@sambego/storybook-state';
import OptionsHelper from '../../utils/helpers/options-helper';
import { Sidebar, SidebarHeader } from './sidebar';
import Button from '../button';
import notes from './notes.md';

const store = new Store({
  open: false
});

const onCancel = () => {
  store.set({ open: false });
  action('cancel')();
};

const openSidebar = () => {
  store.set({ open: true });
  action('open')();
};


storiesOf('Sidebar', module)
  .add('default', () => {
    const enableBackgroundUI = boolean('enableBackgroundUI', false);
    const position = select('position', OptionsHelper.alignBinary[0]);
    const size = select('size', OptionsHelper.sizesFull[3]);
    const className = text('className', '');

    return (
      <State store={ store }>
        <Button onClick={ openSidebar }>Open sidebar</Button>
        <Sidebar
          enableBackgroundUI={ enableBackgroundUI }
          open={ store.get('open') }
          position={ position }
          size={ size }
          onCancel={ onCancel }
        >
          <SidebarHeader className={ className }>
            Header Content
          </SidebarHeader>
        Main Content
        </Sidebar>
      </State>
    );
  }, {
    notes: { markdown: notes }
  });
