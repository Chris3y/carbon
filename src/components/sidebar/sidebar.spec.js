import React from 'react';
import 'jest-styled-components';
import { shallow, mount } from 'enzyme';
import Sidebar from './sidebar.component';
import Textbox from '../textbox/textbox';
import { SidebarStyle } from './sidebar.style';
import { assertStyleMatch } from '../../__spec_helper__/test-utils';

describe('Sidebar', () => {
  let wrapper, spy;

  beforeEach(() => {
    spy = jasmine.createSpy();

    wrapper = shallow(
      <Sidebar
        open
        title='Test'
        className='custom-class'
        data-role='baz'
        data-element='bar'
        onCancel={ spy }
      >
        <Textbox />
        <Textbox />
        <Textbox />
      </Sidebar>
    );
  });

  describe('sidebarClasses', () => {
    it('returns a base sidebar', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('render', () => {
    describe('when sidebar is closed', () => {
      it('sets all the correct classes', () => {
        wrapper = mount(<Sidebar onCancel={ spy } />);
        expect(wrapper.find('.carbon-sidebar').text()).toEqual('');
      });
    });

    describe('when enableBackgroundUI is enabled', () => {
      it('sets all the correct classes', () => {
        wrapper = shallow(
          <Sidebar
            open
            enableBackgroundUI
            size='small'
            position='left'
            onCancel={ spy }
          />
        );
        expect(wrapper.find('.carbon-modal__background').length).toEqual(0);
      });
    });

    describe('when there is no onCancel prop', () => {
      it('should not have a close button', () => {
        wrapper = shallow(
          <Sidebar
            open
          />
        );
        expect(wrapper.find('.carbon-sidebar__close').length).toEqual(0);
      });
    });
  });

  describe('Behaviour', () => {
    describe('clicking the close icon sidebar', () => {
      it('closes the sidebar', () => {
        const icon = wrapper.find('.carbon-sidebar__close-icon');
        icon.simulate('click');
        expect(spy).toHaveBeenCalled();
      });
    });
  });
});

describe('SidebarStyle', () => {
  describe('when prop size is passed to the component and position is set to right', () => {
    const wrapper = mount(<SidebarStyle
      open size='extra-small'
      position='right'
    />);

    it('should render correct style', () => {
      assertStyleMatch({
        width: '150px',
        boxShadow: '-10px 0 15px rgba(0,0,0,0.05)',
        borderLeft: '1px solid #ccd6db',
        right: '0'
      }, wrapper);
    });
  });

  describe('when prop left is passed to the component', () => {
    const wrapper = mount(<SidebarStyle
      open size='extra-small'
      position='left'
    />);

    it('should render correct style', () => {
      assertStyleMatch({
        boxShadow: '10px 0 15px rgba(0,0,0,0.05)',
        borderRight: '1px solid #ccd6db',
        left: '0'
      }, wrapper);
    });
  });
});
