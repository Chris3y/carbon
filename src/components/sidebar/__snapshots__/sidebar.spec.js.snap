// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar sidebarClasses returns a base sidebar 1`] = `
<Portal
  key="1"
>
  <div
    className="carbon-sidebar custom-class"
    data-component="sidebar"
    data-element="bar"
    data-role="baz"
    data-state="open"
  >
    <CSSTransitionGroup
      component="div"
      transitionAppear={true}
      transitionAppearTimeout={500}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="modal-background"
    >
      <div
        className="carbon-modal__background"
      />
    </CSSTransitionGroup>
    <CSSTransitionGroup
      component="div"
      transitionAppear={true}
      transitionAppearTimeout={500}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="sidebar--right"
    >
      <styled.div
        data-component="sidebar"
        data-element="bar"
        data-role="baz"
        position="right"
        size="medium"
      >
        <span
          className="carbon-sidebar__close"
        >
          <Icon
            bgSize="small"
            className="carbon-sidebar__close-icon"
            data-element="close"
            onClick={[Function]}
            type="close"
          />
        </span>
        <Textbox />
        <Textbox />
        <Textbox />
      </styled.div>
    </CSSTransitionGroup>
  </div>
</Portal>
`;
